pkg/apiserver/clusterinfo/service.go

```diff
@@ -0,0 +1,172 @@
+)
+
+type ClusterInfo struct {
+	TiDB         []clusterinfo.TiDB       `json:"tidb"`
+	TiKV         []clusterinfo.TiKV       `json:"tikv"`
+	Pd           []clusterinfo.PD         `json:"pd"`
+	Grafana      clusterinfo.Grafana      `json:"grafana"`
+	AlertManager clusterinfo.AlertManager `json:"alert_manager"`
+}
+
+type Service struct {
+	config  *config.Config
+	etcdCli *etcdclientv3.Client
+	pdCli   pdclient.Client
+}
+
+func NewService(config *config.Config, pdClient pdclient.Client, etcdClient *etcdclientv3.Client) *Service {
+	return &Service{etcdCli: etcdClient, config: config, pdCli: pdClient}
+}
+
+func (s *Service) Register(r *gin.RouterGroup, auth *user.AuthService) {
+	endpoint := r.Group("/topology")
+	//endpoint.Use(auth.MWAuthRequired())
+	endpoint.GET("/", s.topologyHandler)
+	endpoint.DELETE("/tidb/:address/", s.deleteDBHandler)
+}
+
+// @Summary Delete tidb ns.
+// @Description Delete TiDB with ip:port
+// @Produce json
+// @Success 204  Delete successfully and return 204.
+// @Failure 404  The key doesn't exists.
+// @Router /topology/address [delete]
+func (s *Service) deleteDBHandler(c *gin.Context) {
+	v, exists := c.Params.Get("address")
+	if !exists {
+		c.Status(500)
+		_ = c.Error(fmt.Errorf("address not exists in path"))
+		return
+	}
+	address := v
+	ttlKey := fmt.Sprintf("/topology/tidb/%v/ttl", address)
+	nonTTLKey := fmt.Sprintf("/topology/tidb/%v/info", address)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+
+	_, err := s.etcdCli.Delete(ctx, nonTTLKey)
+	if err != nil {
+		c.Status(500)
+		_ = c.Error(err)
+		return
+	}
+	_, err = s.etcdCli.Delete(ctx, ttlKey)
+	if err != nil {
+		c.Status(500)
+		_ = c.Error(err)
+		return
+	}
+
+	c.JSON(http.StatusNoContent, nil)
+}
+
+type ResponseWithErr struct {
+	TiDB         interface{} `json:"tidb"`
+	TiKV         interface{} `json:"tikv"`
+	Pd           interface{} `json:"pd"`
+	Grafana      interface{} `json:"grafana"`
+	AlertManager interface{} `json:"alert_manager"`
+}
+
+// @Summary Dashboard info
+// @Description Get information about the dashboard service.
+// @Produce json
+// @Success 200 {object} ClusterInfo
+// @Router /topology [get]
+func (s *Service) topologyHandler(c *gin.Context) {
+	var info ClusterInfo
+	var returnObject ResponseWithErr
+	errMap := map[string]error{}
+	var errMutex sync.Mutex
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	fetchers := []fetcher{
+		tidbFetcher{},
+		tikvFetcher{},
+		pdFetcher{},
+	}
+
+	var wg sync.WaitGroup
+	for _, fetcher := range fetchers {
+		wg.Add(1)
+		currentFetcher := fetcher
+		go func() {
+			defer wg.Done()
+			err := currentFetcher.fetch(ctx, &info, s)
+			if err != nil {
+				errMutex.Lock()
+				errMap[currentFetcher.name()] = err
+				errMutex.Unlock()
+			}
+		}()
+	}
+	wg.Wait()
+	var exists bool
+	var err error
+
+	if err, exists = errMap["tikv"]; exists {
```

Comment:
I think you should handle these logic in your fetcher. Otherwise there is no way to know the topology from the backend.